keras_nlp/models/llama/llama_attention.py

```diff
@@ -0,0 +1,194 @@
+        self.bias_initializer = keras.initializers.get(bias_initializer)
+        self.max_sequence_length = max_sequence_length
+
+        self.rotary_dim = int(self.attn_head_size * rotary_percentage)
+        self.rope_scaling_type = rope_scaling_type
+        self.rope_scaling_factor = rope_scaling_factor
+
+        if rope_scaling_type is None:
+            self.rotary_embedding_layer = RotaryEmbedding(
+                max_wavelength=rotary_max_wavelength
+            )
+        else:
+            if rope_scaling_type == "linear":
+                self.rotary_embedding_layer = RotaryEmbedding(
+                    max_wavelength=rotary_max_wavelength
+                )
+            elif rope_scaling_type == "dynamic":
+                self.rotary_embedding_layer = DynamicNTKScaledRotaryEmbedding(
+                    max_wavelength=rotary_max_wavelength
+                )
+            else:
+                raise ValueError(
+                    f"Unknown RoPE scaling type {rope_scaling_type}"
+                )
+
+    def build(self, input_shape):
+        self._query_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_heads * self.attn_head_size),
+            name="query",
+        )
+        self._query_dense.build(input_shape)
+
+        self._key_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_key_value_heads * self.attn_head_size),
+            name="key",
+        )
+        self._key_dense.build(input_shape)
+
+        self._value_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_key_value_heads * self.attn_head_size),
+            name="value",
+        )
+        self._value_dense.build(input_shape)
+
+        self._softmax = keras.layers.Softmax(axis=-1, name="attention_softmax")
+
+        self._output_dense = keras.layers.EinsumDense(
+            equation="abc,cd->abd",
+            output_shape=(None, self.hidden_dim),
+            name="attention_output",
+        )
+        self._output_dense.build(input_shape)
+
+        self.built = True
+
+    def call(self, hidden_states, attention_mask=None):
+        query = self._query_dense(hidden_states)
+        key = self._key_dense(hidden_states)
+        value = self._value_dense(hidden_states)
+
+        query_rot, query_pass = (
+            query[..., : self.rotary_dim],
+            query[..., self.rotary_dim :],
+        )
+        key_rot, key_pass = (
+            key[..., : self.rotary_dim],
+            key[..., self.rotary_dim :],
+        )
+
+        query_rot = self.rotary_embedding_layer(query_rot)
+        key_rot = self.rotary_embedding_layer(key_rot)
+
+        query = ops.concatenate((query_rot, query_pass), axis=-1)
+        key = ops.concatenate((key_rot, key_pass), axis=-1)
+
+        key = self._repeat_kv(key, self.num_key_value_groups)
```

Comment:
@mattdangerw can we reuse https://keras.io/api/layers/reshaping_layers/repeat_vector/ ?

Reply:
I think probably simplest to stick with `ops` here. Would `ops.tile(key, [1, 1, self.num_key_value_groups, 1])` work by any chance?

Reply:
Thanks, I didn't know about this thing exist !